crates/rethnet_evm/benches/state_debug.rs

```diff
@@ -0,0 +1,330 @@
+            debug_assert!(result.is_ok());
+            debug_assert!(result.unwrap().is_none());
+        },
+    );
+}
+
+fn bench_account_storage_root_account_exists(c: &mut Criterion) {
+    bench_sync_state_method(
+        c,
+        "StateDebug::account_storage_root(), account exists",
+        prep_no_op,
+        |state, number_of_accounts| {
+            let address = Address::from_low_u64_ne(number_of_accounts);
+            debug_assert!(state.basic(address).unwrap().is_some());
+            let result = state.account_storage_root(&address);
+            debug_assert!(result.is_ok());
+            debug_assert!(result.unwrap().is_some());
+        },
+    );
+}
+
+fn bench_insert_account_already_exists(c: &mut Criterion) {
+    bench_sync_state_method(
+        c,
+        "StateDebug::insert_account(), account already exists",
+        prep_no_op,
+        |mut state, number_of_accounts| {
+            let address = Address::from_low_u64_ne(number_of_accounts);
+            debug_assert!(state.basic(address).unwrap().is_some());
+            let result = state.insert_account(address, AccountInfo::default());
+            debug_assert!(result.is_ok())
+        },
+    );
+}
+
+fn bench_insert_account_doesnt_exist(c: &mut Criterion) {
+    bench_sync_state_method(
+        c,
+        "StateDebug::insert_account(), account doesn't yet exist",
+        prep_no_op,
+        |mut state, number_of_accounts| {
+            let address = Address::from_low_u64_ne(number_of_accounts + 1);
+            debug_assert!(state.basic(address).unwrap().is_none());
+            let result = state.insert_account(address, AccountInfo::default());
+            debug_assert!(result.is_ok());
+        },
+    );
+}
+
+fn bench_modify_account_doesnt_exist(c: &mut Criterion) {
+    bench_sync_state_method(
+        c,
+        "StateDebug::modify_account(), account doesn't exist",
+        prep_no_op,
+        |mut state, number_of_accounts| {
+            let address = Address::from_low_u64_ne(number_of_accounts + 1);
+            debug_assert!(state.basic(address).unwrap().is_none());
+            let result = state.modify_account(
+                address,
+                AccountModifierFn::new(Box::new(|_balance, &mut _nonce, _code| {})),
+                &|| Ok(AccountInfo::default()),
+            );
+            debug_assert!(result.is_ok());
+        },
+    );
+}
+
+fn bench_modify_account_exists_with_code_no_change(c: &mut Criterion) {
+    bench_sync_state_method(
+        c,
+        "StateDebug::modify_account(), account already exists, with code, no code change",
+        prep_no_op,
+        |mut state, number_of_accounts| {
+            let address = Address::from_low_u64_ne(number_of_accounts);
+            debug_assert!(state.basic(address).unwrap().is_some());
+            // TODO: figure out why the following assert is failing
+            //debug_assert!(state.basic(address).unwrap().unwrap().code.is_some());
```

Comment:
> Code is split from the account itself and stored in a separate data structure.

But, why? Doesn't that seem like an implementation detail? Why have a `code` field that we just ignore like that? I don't presume to know enough to advocate for one way or the other, but ignoring a field in an output struct feels like a pretty counter-intuitive interface to me.

I went on a bit of a rabbit hole "fixing" this before seeing your comment. See my draft PR [here](https://github.com/NomicFoundation/hardhat/pull/3918). It seems like a pretty straightforward way to support the interface as expected. What do you think?

Reply:
It's not a struct we control. It's inside `revm`. I agree with you though: code and code_hash should be combined in an enum to avoid having empty fields.

It's documented though: https://docs.rs/revm-primitives/latest/revm_primitives/state/struct.AccountInfo.html